CASMCMS-7676 Add documentation for IMS recipe templating (#974)
Cray-HPE/docs-csm: operations/image_management/Upload_and_Register_an_Image_Recipe.md

    type = "s3"
    ```

## Templating IMS Recipes when Building an IMS Image

IMS can optionally template the contents of an IMS recipe before building the recipe with Kiwi-NG during an IMS create job. This 
enables variables stored in the recipe to be dynamically replaced with values registered with the IMS recipe record after the 
recipe is downloaded from S3, but before it is built by Kiwi-NG. This functionality is used by various Cray products to ensure 
that the resulting image is built from the most correct release versions of Nexus hosted RPM repositories. 

Follow the steps below to enable IMS templating within a given recipe. Note that for IMS to properly template a recipe,
the following procedures must all be completed.

### Enable Templating in IMS Recipe Archive tgz File

> 1. Add a file named '.ims_recipe_template.yaml' to the root of the recipe archive. 

Reviewer: ```suggestion
1. Add a file named `.ims_recipe_template.yaml` to the root of the recipe archive. 
```

   ``` bash
   ncn# ls -al
   ```
   
   Example output:

>    ``` bash

Reviewer: ```suggestion
   ```
```

   drwxr-xr-x 3 root root    70 Feb 22 19:05 .
   drwxr-xr-x 3 root root    41 Feb 22 17:18 ..      
   -rw-r--r-- 1 root root  2744 Feb  4 21:14 config.sh
   -rw-r--r-- 1 root root 24645 Feb  4 21:15 config.xml
   -rw-r--r-- 1 root root  1216 Feb  4 21:14 images.sh
   -rw-r--r-- 1 root root    29 Feb  4 21:14 .ims_recipe_template.yaml
   drwxr-xr-x 3 root root    18 Feb  4 21:15 root
   ```
   

>    The contents of this file must be valid yaml containing the keyword `template_files`. The `template_files` keyword 

Reviewer: ```suggestion
   The contents of this file must be valid YAML, containing the keyword `template_files`. The `template_files` keyword
```

   must contain a list of files to be templated within the image recipe archive.

   ``` bash
   ncn# cat .ims_recipe_template.yaml
   ```

   Example output:   

>    ```

Reviewer: ```suggestion
   ```yaml
```

   template_files:
   - config.xml
   ```
   
   IMS will template the indicated files 'in place' -- that is, the listed file will be replaced with the templated copy.

>    **NOTE:** Only files listed under the template_files key will be templated by IMS. 

Reviewer: ```suggestion
   **NOTE:** Only files listed under the `template_files` key will be templated by IMS. 
```

2. For each indicated file above, add appropriate Jinja2 variables where required to enable the result that are sought.
   In the example below, the Jinja2 variable `{{ CSM_RELEASE_VERSION }}` is used to help identify the fully qualified 
   Nexus repo name/path to use when building the recipe.

   ```bash
   ncn# cat config.xml
   ...
   <repository type="rpm-md" alias="csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3" priority="2" imageinclude="true">
       <source path="https://packages.local/repository/csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3/"/>
   </repository>
   ...
   ```

   **NOTE:** The repo referenced above is for documentation purposes only and may not actually exist. 

> 3. Create a tgz archive of the image recipe.

Reviewer: ```suggestion
3. Create a `tgz` archive of the image recipe.
```

    ```bash
    ncn# export ARTIFACT_FILENAME=recipe.tar.gz
    ncn# tar cvfz ../$ARTIFACT_FILE .
    ncn# cd ..
    ```

### Add Template Key/Value Pairs to IMS Recipe Record

1. Create a new IMS recipe record with template_dictionary key/value pairs.

    ```bash
    ncn# cray ims recipes create --name "My Recipe" \
    --recipe-type kiwi-ng --linux-distribution sles15 \
    --template-dictionary-key CSM_RELEASE_VERSION \
    --template-dictionary-value 1.2.5
    ```
   
    Example output:
    

>     ``` bash

Reviewer: ```suggestion
    ```
```

    created = "2018-12-04T17:25:52.482514+00:00"
    id = "2233c82a-5081-4f67-bec4-4b59a60017a6"
    linux_distribution = "sles15"
    name = "my_recipe.tgz"
    recipe_type = "kiwi-ng"
    
    [[template_dictionary]]
    key = "CSM_RELEASE_VERSION"
    value = "1.2.5"
    ```

    Additional key/value pairs can be added by providing a list of comma-separated keys/values to the 
    `--template-dictionary-key` and `--template-dictionary-value` parameters.  

    If successful, create a variable for the `id` value in the returned data.

    ```bash
    ncn# export IMS_RECIPE_ID=2233c82a-5081-4f67-bec4-4b59a60017a6
    ```
   
2. Upload the customized recipe to S3.

   It is suggested as a best practice that the S3 object name start with `recipes/` and contain the IMS recipe ID to remove ambiguity.

   ```bash
   ncn# cray artifacts create ims recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME $ARTIFACT_FILENAME
   ```

3. Update the IMS recipe record with the S3 path to the recipe archive.

   ```bash
   ncn# cray ims recipes update $IMS_RECIPE_ID \

>    --link-type s3 \

Reviewer: ```suggestion
           --link-type s3 \
```

>    --link-path s3://ims/recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME

Reviewer: ```suggestion
           --link-path s3://ims/recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME
```

   ```

   Example output:

>    ``` bash

Reviewer: ```suggestion
   ```
```

   id = "2233c82a-5081-4f67-bec4-4b59a60017a6"
   recipe_type = "kiwi-ng"
   linux_distribution = "sles15"
   name = "my_recipe.tgz"
   created = "2020-02-05T19:24:22.621448+00:00"

   [[template_dictionary]]
   key = "CSM_RELEASE_VERSION"
   value = "1.2.5"
   
   [link]
   path = "s3://ims/recipes/2233c82a-5081-4f67-bec4-4b59a60017a6/recipe.tar.gz"
   etag = ""
   type = "s3"
   ```

### Build an Image from an IMS Templated Recipe

The procedure to build an image from an IMS recipe that uses templating does not change. Follow the normal IMS create
procedure, specifying the recipe's ID value in the job's `--artifact-id` parameter. The IMS job will start as normal,
but after downloading the recipe from S3, there will be an indication that IMS is templating the recipe in the job's
`fetch-recipe` container.

```

> ncn# i logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe

Reviewer 2: Use full commands in docs, not shortcuts.

> ncn# i logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe

Reviewer: ```suggestion
ncn# kubectl -n ims logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe
```

```

Example Output:

> ``` bash

Reviewer: ```suggestion
```
```
